subworkflows/local/cooler.nf

```diff
@@ -0,0 +1,56 @@
+/*
+ * pair the proper mapped pairs
+ */
+params.options = [:]
+
+include { COOLER_CLOAD   } from '../../modules/local/cooler/cload/main'   addParams(options: params.options.cooler_cload)
+include { COOLER_MERGE   } from '../../modules/local/cooler/merge/main'   addParams(options: params.options.cooler_merge)
+include { COOLER_ZOOMIFY } from '../../modules/local/cooler/zoomify/main' addParams(options: params.options.cooler_zoomify)
+include { COOLER_DUMP    } from '../../modules/nf-core/modules/cooler/dump/main'    addParams(options: params.options.cooler_dump_per_group)
+include { COOLER_DUMP
+    as COOLER_DUMP_SAMPLE} from '../../modules/nf-core/modules/cooler/dump/main'    addParams(options: params.options.cooler_dump_per_sample)
+include { DUMPINTRAREADS } from '../../modules/local/cooler/dumpintrareads'    addParams(options: params.options.dumpintrareads_per_group)
+include { DUMPINTRAREADS
+    as DUMPINTRAREADS_SAMPLE} from '../../modules/local/cooler/dumpintrareads'    addParams(options: params.options.dumpintrareads_per_sample)
+include { JUICER         } from '../../modules/local/cooler/juicer'    addParams(options: params.options.juicer)
+
+workflow COOLER {
+    take:
+    valid_pairs               // channel: [ val(meta), val(bin), [pairs], [pairs.px] ]
+    chromsizes                // channel: [ path(chromsizes) ]
+    juicer_jvm_params         // values
+    juicer_tools_jar          // channel: [ path(juicer_tool jar) ]
+
+    main:
+    // HiC-like contact matrix
+    ch_version = COOLER_CLOAD(valid_pairs.map{[it[0], it[2], it[3]]}, valid_pairs.map{it[1]}, chromsizes).version
+    // Merge contacts
+    COOLER_CLOAD.out.cool
+                .map{
+                    meta, bin, cool ->
+                    [meta.group, bin, cool]
+                }
+                .groupTuple(by:[0, 1])
+                .map{group, bin, cool -> [[id:group, bin:bin], cool]}
+                .set{ch_cooler}
+    COOLER_MERGE(ch_cooler)
+    // create mcooler file for visualization
+    COOLER_ZOOMIFY(COOLER_MERGE.out.cool)
+    // dump long.intra.bedpe for each group for MAPS to call peaks
+    COOLER_DUMP(COOLER_MERGE.out.cool).bedpe | DUMPINTRAREADS
+    if(juicer_tools_jar){
+        JUICER(DUMPINTRAREADS.out.gi.combine(juicer_tools_jar), chromsizes, juicer_jvm_params)
```

Comment:
Why do you `combine` the jar file here with the input?

Reply:
Indeed I am not sure why. If I do not combine them, only the first one will be run. With `combine` all `gi` will be run.

Reply:
Ah, nextflow has two types of channels. Queue type and value type channels. If input comes in as a queue type channel, then the each element of the queue is consumed and another process cannot reuse it. You need to specify the juicer_tools_jar as a value type channel so it is reused. 
Specifically this line in `hicar.nf`:
```
ch_juicer_tools        = params.juicer_tools_jar ? Channel.fromPath(params.juicer_tools_jar, checkIfExists: true) : Channel.empty()
```
Makes a queue type channel `fromPath`.
I think you can do:
```
ch_juicer_tools        = params.juicer_tools_jar ? file(params.juicer_tools_jar, checkIfExists: true) : Channel.empty()
```
to make it a value type channel (test with a toy example, otherwise the `collect` operator also makes a value type channel).

Reply:
Thank you for explain this. It puzzled me for long time. I will change it to value channel.